### PR TITLE
TS-4914: Fix response headers on 304 response

### PR DESCRIPTION
**Pull request by shinya1020**
https://issues.apache.org/jira/browse/TS-4914

### CODE REVIEW

**Comment by masaori335** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/908/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/803/ for details.

**Comment by zwoop** [comment]
Mostly out of curiosity, is there a particular header here in the (limited) set of headers that is likely to have multi-values?

**Comment by shinya1020** [comment]
In our production, some servers return multiple Cache-Control headers.

For example, if the following configurations are applied to apache httpd
```
Header always set Cache-Control "public"
ExpiresActive On
ExpiresDefault "access plus 1 days"
```
then the server returns the following headers.
```
Cache-Control: public
Cache-Control: max-age=86400
Expires: Tue, 04 Oct 2016 12:25:44 GMT
```

File: proxy/http/HttpTransact.cc [hunk 1]
```diff
@@ -7948,9 +7948,11 @@ HttpTransact::build_response(State *s, HTTPHdr *base_response, HTTPHdr *outgoing
               const char *value;
 
               field = base_response->field_find(fields[i].name, fields[i].len);
-              ink_assert(field != nullptr);
-              value = field->value_get(&len);
-              outgoing_response->value_append(fields[i].name, fields[i].len, value, len, 0);
+              while (field) {
+                value = field->value_get(&len);
+                outgoing_response->value_append(field_name[i], field_len[i], value, len, true);
```
Comment (maskit) [inline proxy/http/HttpTransact.cc]: This fix can be applicable for Cache-Control header and some other headers but not for every headers.
Please refer to the Note in [RFC7230 Section 3.2.2](https://tools.ietf.org/html/rfc7230#section-3.2.2). At least it's not applicable for Set-Cookie header.

We might be able to have a blacklist / whitelist, but it won't be perfect list. Maybe, just sending all headers separately is a safer way?

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
Sounds to me we should sent both the header lines to the client instead of sending only the first one.

**Comment by zwoop** [comment]
I agree. If the origin is misbehaving, and sends e.g. duplicate set-cookie headers, is it really our responsibility to "fix" broken origins?

File: proxy/http/HttpTransact.cc (hunk 1 above)
Comment (bryancall) [inline proxy/http/HttpTransact.cc]: @maskit On this we are not looping over the Set-Cookie header.  We are looping over the 5 headers defined by fields[].  Can you please review again?

**Comment by bryancall** [comment]
@shinya1020  Can you please update the PR?  We would like to land this.

File: proxy/http/HttpTransact.cc (hunk 1 above)
Comment (maskit) [inline proxy/http/HttpTransact.cc]: Sorry, I didn't see the header list. Seems OK then.

**Comment by zwoop** [comment]
@maskit I think I agree. If there are header duplications, it's not our job to "optimize" by merging; we should simply preserve the duplication, and copy all entries for that same header.

**Comment by bryancall** [comment]
@shinya1020 Please make the change to add multiple headers when there are duplicate headers from the origin.

**Comment by bryancall** [comment]
Please reopen when the review concerns have been addressed.

**Pull request closed without merging**